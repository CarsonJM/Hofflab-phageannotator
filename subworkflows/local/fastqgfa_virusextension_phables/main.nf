//
// Extend assembly graphs using Phables
//

include { PHABLES_INSTALL   } from '../../../modules/local/phables/install/main'
include { PHABLES_RUN       } from '../../../modules/local/phables/run/main'

workflow FASTQGFA_VIRUSEXTENSION_PHABLES {
    take:
    fastq_gz            // [ [ meta ], [ reads_1.fastq.gz, reads_2.fastq.gz ] ] , read files (mandatory)
    gfa_gz              // [ [ meta ], gfa.gz ]                                 , assembly graph (mandatory)
    phables_db          // [ phables_db ]                                       , Phables database (mandatory)

    main:
    ch_versions = Channel.empty()

    // if genomad_db exists, skip PHABLES_INSTALL
    if (phables_db) {
        ch_phables_db = phables_db
    } else {
        //
        // MODULE: download geNomad database
        //
        ch_phables_db   = PHABLES_INSTALL( ).phables_db
        ch_versions     = ch_versions.mix(PHABLES_INSTALL.out.versions)
    }

    // combine fastq and gfa for phables
    ch_phables_input = fastq_gz
        .join(gfa_gz)
        .multiMap { meta, fastq, gfa ->
            meta.assembler = meta.assembler + "_phables"
            fastq : [ meta, fastq ]
            gfa   : [ meta, gfa ]
        }

    //
    // MODULE: Run phables to extend assemblies
    //
    PHABLES_RUN(
        ch_phables_input.gfa,
        ch_phables_input.fastq,
        ch_phables_db.first()
    )
    ch_phables_fasta_gz = PHABLES_RUN.out.fasta
    ch_versions = ch_versions.mix(PHABLES_RUN.out.versions)


    emit:
    phables_fasta   = ch_phables_fasta_gz   // [ [ meta ], extended_contigs.fna.gz ]    , FASTA file containing extended contigs
    versions        = ch_versions.unique()  // [ versions.yml ]
}
