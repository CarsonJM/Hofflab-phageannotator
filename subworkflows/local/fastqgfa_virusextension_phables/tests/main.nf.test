nextflow_workflow {

    name "Test Subworkflow: FASTQGFA_VIRUSEXTENSION_PHABLES"
    script "subworkflows/local/fastqgfa_virusextension_phables/main.nf"
    workflow "FASTQGFA_VIRUSEXTENSION_PHABLES"

    setup {
        run("PHABLES_INSTALL") {
            script "../../../../modules/local/phables/install/main.nf"
        }
    }

    test("fastq.gz + gfa + no db") {
        when {
            workflow {
                """
                input[0] = Channel.of(
                    [
                        [ id: 'test1' ],
                        [
                            file("${projectDir}/assets/test-datasets/fastqs/test1_1.fastq.gz", checkIfExists: true),
                            file("${projectDir}/assets/test-datasets/fastqs/test1_2.fastq.gz", checkIfExists: true)
                        ]
                    ],
                    [
                        [ id: 'test2' ],
                        [
                            file("${projectDir}/assets/test-datasets/fastqs/test1_1.fastq.gz", checkIfExists: true),
                            file("${projectDir}/assets/test-datasets/fastqs/test1_2.fastq.gz", checkIfExists: true)
                        ]
                    ]
                )
                input[1] = Channel.of(
                    [
                        [ id: 'test1' ],
                        [
                            file("${projectDir}/assets/test-datasets/gfa/test1.assembly.gfa", checkIfExists: true),
                        ]
                    ],
                    [
                        [ id: 'test2' ],
                        [
                            file("${projectDir}/assets/test-datasets/gfa/test1.assembly.gfa", checkIfExists: true),
                        ]
                    ]
                )
                input[2] = null
                """
            }
        }
        then {
            assertAll (
                { assert workflow.success },
                { assert snapshot(workflow.out.versions).match() },
                { assert new File(workflow.out.phables_fasta[0][1]).exists() }
            )
        }
    }


    test("fastq.gz + gfa + db") {
        when {
            workflow {
                """
                input[0] = Channel.of(
                    [
                        [ id: 'test1' ],
                        [
                            file("${projectDir}/assets/test-datasets/fastqs/test1_1.fastq.gz", checkIfExists: true),
                            file("${projectDir}/assets/test-datasets/fastqs/test1_2.fastq.gz", checkIfExists: true)
                        ]
                    ],
                    [
                        [ id: 'test2' ],
                        [
                            file("${projectDir}/assets/test-datasets/fastqs/test1_1.fastq.gz", checkIfExists: true),
                            file("${projectDir}/assets/test-datasets/fastqs/test1_2.fastq.gz", checkIfExists: true)
                        ]
                    ]
                )
                input[1] = Channel.of(
                    [
                        [ id: 'test1' ],
                        [
                            file("${projectDir}/assets/test-datasets/gfa/test1.assembly.gfa", checkIfExists: true),
                        ]
                    ],
                    [
                        [ id: 'test2' ],
                        [
                            file("${projectDir}/assets/test-datasets/gfa/test1.assembly.gfa", checkIfExists: true),
                        ]
                    ]
                )
                input[2] = null
                """
            }
        }
        then {
            assertAll (
                { assert workflow.success },
                { assert snapshot(workflow.out).match() },
                { assert new File(workflow.out.phables_fasta[0][1]).exists() }
            )
        }
    }
}
