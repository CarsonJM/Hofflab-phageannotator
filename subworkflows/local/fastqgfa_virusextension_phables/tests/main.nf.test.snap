{
    "fastq.gz + gfa + no db": {
        "content": [
            [
                "versions.yml:md5,47c9014d30b6d0dbd2912ceba0cee284",
                "versions.yml:md5,47c9014d30b6d0dbd2912ceba0cee284",
                "versions.yml:md5,f66a187fec28951dd902ad63bbe5a077"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.1"
        },
        "timestamp": "2024-07-24T11:22:11.043199799"
    },
    "fastq.gz + gfa + db": {
        "content": [
            [
                "versions.yml:md5,47c9014d30b6d0dbd2912ceba0cee284",
                "versions.yml:md5,47c9014d30b6d0dbd2912ceba0cee284",
                "versions.yml:md5,f66a187fec28951dd902ad63bbe5a077"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.1"
        },
        "timestamp": "2024-07-24T11:50:27.627194312"
    }
}
